sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/PredicateSuite.scala

```diff
@@ -35,7 +35,8 @@ class PredicateSuite extends SparkFunSuite with ExpressionEvalHelper {
     test(s"3VL $name") {
       truthTable.foreach {
         case (l, r, answer) =>
-          val expr = op(NonFoldableLiteral(l, BooleanType), NonFoldableLiteral(r, BooleanType))
+          val expr = op(NonFoldableLiteral.create(l, BooleanType),
```

Comment:
The problem is that NonFoldableLiteral(java.math.BigDecimal, DecimalType) is keeping the java value while NonFoldableLiteral.create(_,_) does Literal.create which transforms java.math.BigDecimal to Decimal (catalyst type).
It's a bit of weird code. Literal has the same problem. Literal(_,_) is not the same as Literal.create(_,_). So, one should always use .create and never the 2 parameter constructor.
Otherwise you get java values in the execution, which in my case was failing because there is no ordering for non catalyst types.